go-controller/pkg/node/gateway_localnet.go

```diff
@@ -189,47 +205,41 @@ func initLocalnetGateway(nodeName string, subnet *net.IPNet, wf *factory.WatchFa
 		ChassisID:      chassisID,
 		InterfaceID:    ifaceID,
 		MACAddress:     macAddress,
-		IPAddresses:    []*net.IPNet{gatewayIPCIDR},
-		NextHops:       []net.IP{gatewayNextHop},
+		IPAddresses:    gatewayIPs,
+		NextHops:       nextHops,
 		NodePortEnable: config.Gateway.NodeportEnable,
 	})
 	if err != nil {
 		return err
 	}
 
-	if utilnet.IsIPv6CIDR(subnet) {
-		// TODO - IPv6 hack ... for some reason neighbor discovery isn't working here, so hard code a
-		// MAC binding for the gateway IP address for now - need to debug this further
-		err = util.LinkNeighAdd(link, gatewayIP, macAddress)
-		if err == nil {
-			klog.Infof("Added MAC binding for %s on %s", gatewayIP, localnetGatewayNextHopPort)
-		} else {
-			klog.Errorf("Error in adding MAC binding for %s on %s: %v", gatewayIP, localnetGatewayNextHopPort, err)
+	for _, ip := range gatewayIPs {
+		ipt, err := localnetIPTablesHelper(ip.IP)
+		if err != nil {
+			return err
 		}
-	}
 
-	ipt, err := localnetIPTablesHelper(subnet)
-	if err != nil {
-		return err
-	}
-
-	err = localnetGatewayNAT(ipt, localnetGatewayNextHopPort, gatewayIP)
-	if err != nil {
-		return fmt.Errorf("failed to add NAT rules for localnet gateway (%v)", err)
-	}
+		err = localnetGatewayNAT(ipt, localnetGatewayNextHopPort, ip.IP)
+		if err != nil {
+			return fmt.Errorf("failed to add NAT rules for localnet gateway (%v)", err)
+		}
 
-	if config.Gateway.NodeportEnable {
-		err = localnetNodePortWatcher(ipt, wf, gatewayIP)
+		if config.Gateway.NodeportEnable {
+			err = localnetNodePortWatcher(ipt, wf, ip.IP)
```

Comment:
I thought we beed to install IP table rules for node port services based on the IP family of cluster IP even in case of dual stack as services are essentially single stack.  Is that not needed ? This code calls  `localnetNodePortWatcher` twice for dual stack and hence registers the call back twice for node port services and installs both IP table rules. Please correct me if my understanding is not correct.

Reply:
Well, Services are becoming potentially-dual-stack in kube 1.19. But either way, yes, you're right that we should check `svc.ipFamily` for each service we process, and not install IPv4 rules for IPv6-only services or IPv6 rules for IPv4-only services

Reply:
pushed a fixup...